### DISCUSSION TITLE
Loading a third-party services Script/Widget NOT working App_Router

### DISCUSSION THREAD

**Discussion by operator-axel** [vercel/next.js#64042] [Help]
### Summary

I'm sorry, I'm new and learning but built a website for my buddy's company and now - he wants to add these services third party widgets, and it's a mess. So after now days scouring every source (yes I'm sure there is an answer out there, but not one I could comprehend with my limited scope). 

After many trial and error their support responded with this: 

> Response from our tech team: Our widgets are **rendered at the moment the DOMContentLoaded event occurs**, and if the client renders our scripts after this event occurs, then the script is no longer executed and our widget does not appear in the container. Possible solutions strongly depend on the internal architecture of the client's website. The client's web-developers must understand whether their architecture allows them to bypass this or not, using the information provided above.

And I'm postponing attaching all the different iterations of failure to keep from cluttering.

These are the snippets I need to import:

Thank you in advance, and forgive my ignorance. I was just learning how to code and it turned into this. 

```
`<script src="https://www.enhancify.com/realwidget/?page=9923018&color1=%23B31417&color2=%23064C8F&color3=%23FFFFFF"></script><div class='realwidget' data-textButton='Apply now' data-widthbtn='400'></div>`
```

```
<script src="https://www.enhancify.com/fullpagewidget/"></script><div id='fullpagewidget' data-color1='#B31417' data-color2='#064C8F' data-coBrandedColor='#FFFFFF' data-page='9923018'></div>
```

**Comment by icyJoseph** [comment]
Have you tried using the `next/script` component, with strategy set to `beforeInteractive`? 

https://nextjs.org/docs/app/building-your-application/optimizing/scripts#strategy

**Comment by operator-axel** [comment]
So from what I took from that documentation and just reread is that is only for site-wide scripts and needs to be placed in the root layout. Or am I misunderstanding? Thank you for the quick reply

> beforeInteractive scripts must be placed inside the root layout (app/layout.tsx) and are designed to load scripts that are needed by the entire site (i.e. the script will load when any page in the application has been loaded server-side).

**Reply by icyJoseph** [reply]
Yeah, but nevermind that now. Does that make it work?

If so, maybe look at, useServerInsertedHTML:

- https://nextjs.org/docs/app/building-your-application/styling/css-in-js
- https://github.com/vercel/next.js/discussions/49354


Or at the React script tag:

- https://react.dev/reference/react-dom/components/script#noun-labs-1201738-(2)
- https://react.dev/reference/react-dom/preinit

**Reply by operator-axel** [reply]
Okay, so I tried it in main layout and Negative - it just made everything laggy like it was just loaded in the background. Also still tried the solutions from the links you laid out - and I got each of them running but to same effect - just the base layout and nothing else..

**Reply by icyJoseph** [reply]
I haven't checked their script, but could you explain further what's expected to happen, and what's actually happening? Are they waiting for the DOMContentLoaded event, and without checking if that's already happened?

**Reply by operator-axel** [reply]
so this is what they said - does this answer your question? 

> rendered at the moment the DOMContentLoaded event occurs, and if the client renders our scripts after this event occurs, then the script is no longer executed and our widget does not appear in the container

**Reply by icyJoseph** [reply]
Aha, so they target a DOM node, that I guess you setup for them, and then they render into that DOM node. 


Later when the user navigates away from that page, said DOM node is removed because it doesn't belong in the next route, and when coming back to the start, you set-up again, the DOM node, but nothing happens, right?

**Reply by operator-axel** [reply]
Right - to my understanding. 

But no I did not set up anything for them - I got three emails with those widgets and was just told copy and paste in... (or you meant setup for them as in the code and I'm an idiot)

**Reply by icyJoseph** [reply]
You're doing fine. Often, integration with widgets in SPAs is tricky...

I have to close shop now but try this, as a first step.

Encapsulate the code they have given you in a client component. Render that in the root layout. 

Use the technique that makes it work every time. I'd guess next/script with before interactive should do it?

Anyway, just make sure you get it to work consistently on first render, no matter the page. 

If that works, then with CSS for now, when the pathname (usePathname) is not the page you want, apply an inline style with display none on the div where the widget renders (it has a className real widget or something like that) ~ otherwise in the path you want the widget to show, don't apply any style.

In other words, keep the widget in all pages, and manually hide it yourself. Does that make it look like it works?

**Reply by operator-axel** [reply]
![240404_13h00m10s_screenshot](https://github.com/vercel/next.js/assets/77085081/e2e764c6-b678-4b3a-8b70-ebf010dfc0e2)
![240404_13h00m56s_screenshot](https://github.com/vercel/next.js/assets/77085081/d301c531-d90a-4aa1-9c91-f47747691cf5)


So I figured I'd share because I could easily be doing your implementation wrong. But this is the the widget script in its own client component and trying to call it in the root layout. 

Still nothing however.

**Comment by operator-axel** [comment]
So this is the code if you just go to the url of one of the widgets - which I have also tried just manually injecting but to no avail but temporary success doing that - it mounts once then disappears on nav or reload. Providing as maybe it will help the people much smarter than me here understand the problem.

```
document.addEventListener("DOMContentLoaded", () => {
    let el = document.querySelectorAll('.realwidget');
    let body = document.querySelector('body');
    el.forEach((elItem) => {
    elItem.innerHTML="<div class='realwidget_button' style='background:#B31417;border: 2px solid #B31417; color: #FFFFFF"; width:calc("+elItem.dataset.widthbtn+"px - 10px);'>"+elItem.dataset.textbutton+"</div>";
    })
            let s = document.createElement("link");
            s.href = "https://www.enhancify.com/build/css/marketplace_realwidget_button.css";
            s.rel = "stylesheet";
            body.append(s);

            let pop_up = document.createElement("div");
            pop_up.className  = "pop_up-realwidget";
            body.append(pop_up);

            let pop_up_section = document.createElement("div");
            pop_up_section.className  = "pop_up-realwidget_section";
            pop_up.append(pop_up_section);

            let pop_up_close = document.createElement("div");
            pop_up_close.className  = "pop_up-realwidget_close";
            pop_up_section.append(pop_up_close);



            el.forEach((elItem) => {
                const buttonClick = elItem.querySelector('.realwidget_button');
                buttonClick.addEventListener('click', function () {
                    if(document.getElementById('realwidget_content')== null){
                        let frame = document.createElement("iframe");
                        let params = 'color1=' + encodeURIComponent('#B31417') +
                        '&color2=' + encodeURIComponent('#064C8F')+'&color3=' + encodeURIComponent('#FFFFFF"');
                        frame.src  =  "https://www.enhancify.com/blue-horizon-exteriors?siteaction=realwidget&utm_source=realwidget&"+params;
                        frame.width="100%";
                        frame.height="95%";
                        frame.frameBorder="0";
                        frame.id="realwidget_content";
                        frame.referrerPolicy="unsafe-url";
                        pop_up_section.append(frame);
                    }
                    pop_up.style.display="flex";
                    let html = document.querySelector('html');
                    html.classList.add('overflow-shadow');
                });
            });


        const closeClick = document.querySelector('.pop_up-realwidget_close');
        pop_up_close.addEventListener('click', function () {
        let html = document.querySelector('html');
            pop_up.style.display="none";
            html.classList.remove('overflow-shadow');
        });
});



```
